Lib/pathlib.py

```diff
@@ -1193,18 +1236,254 @@ def cwd(cls):
         # os.path.abspath('.') == os.getcwd().
         return cls().absolute()
 
+    def expanduser(self):
+        """ Return a new path with expanded ~ and ~user constructs
+        (as returned by os.path.expanduser)
+        """
+        self._unsupported("expanduser")
+
     @classmethod
     def home(cls):
-        """Return a new path pointing to the user's home directory (as
-        returned by os.path.expanduser('~')).
+        """Return a new path pointing to expanduser('~').
         """
         return cls("~").expanduser()
 
+    def readlink(self):
+        """
+        Return the path to which the symbolic link points.
+        """
+        self._unsupported("readlink")
+
+    def resolve(self, strict=False):
+        """
+        Make the path absolute, resolving all symlinks on the way and also
+        normalizing it.
+        """
+        if self._resolving:
+            return self
+        try:
+            path = self.absolute()
+        except UnsupportedOperation:
+            path = self
+
+        def split(path):
+            return path._from_parsed_parts(path.drive, path.root, []), path._tail[::-1]
+
+        link_count = 0
+        stat_cache = {}
+        target_cache = {}
+        path, parts = split(path)
+        while parts:
+            part = parts.pop()
+            if part == '..':
+                if not path._tail:
+                    if path.root:
+                        # Delete '..' segment immediately following root
+                        continue
+                elif path._tail[-1] != '..':
+                    # Delete '..' segment and its predecessor
+                    path = path.parent
+                    continue
+                path = path._make_child_relpath(part)
+            else:
+                lookup_path = path
+                path = path._make_child_relpath(part)
+                path._resolving = True
+                path_str = str(path)
+                try:
+                    st = stat_cache.get(path_str)
+                    if st is None:
+                        st = stat_cache[path_str] = path.stat(follow_symlinks=False)
+                    if S_ISLNK(st.st_mode):
+                        # Like Linux and macOS, raise OSError(errno.ELOOP) if too many symlinks are
+                        # encountered during resolution.
+                        link_count += 1
+                        if link_count >= _MAX_SYMLINKS:
+                            raise OSError(ELOOP, "Too many symbolic links in path", path_str)
+                        target = target_cache.get(path_str)
+                        if target is None:
+                            target = target_cache[path_str] = path.readlink()
```

Comment:
The same point as ``stat_cache``:
```suggestion
                        try:
                            target = target_cache[path_str]
                        except KeyError:
                            target = target_cache[path_str] = path.readlink()
```